### ISSUE TITLE
Create Switch component

### ISSUE THREAD

**Issue by kc4181** [cerner/terra-core#2520]
# Feature Request
Create the Switch component based on the OCS designs posted in zeplin:
https://zpl.io/aRP775E

<img width="291" alt="Screen Shot 2019-12-05 at 2 19 13 PM" src="https://user-images.githubusercontent.com/1355485/70270616-69f72c00-176a-11ea-8333-663bce05aa18.png">

## @ Mentions
@neilpfeiffer

**Comment by neilpfeiffer** [comment]
Will be adding some specific docsite examples to create, but good to start otherwise.

**Comment by nramamurth** [comment]
Update: I have the basic component done on top of native checkbox element and I need to figure out keyboard interaction and I should be able to put out a technical design. 

![switch-toggle](https://user-images.githubusercontent.com/10090859/71578821-b8a4a580-2b1f-11ea-98f6-465d8f29c80a.gif)

**Comment by nramamurth** [comment]
<h3> Switch as a &lt;button> v/s &lt;input type='checkbox'></h3>

---

I started off creating a prototype by implementing `Switch` as a "visually masked" checkbox and I had most of the accessibility concerns addressed except for one - whether or not the role should be "switch". 

While researching further on that topic, I ran into the following post that talks about when it is appropriate to implement a "toggle" switch-like component as a `button` or `input type='checkbox'` with `role='switch'` and the implications of using "switch" as the role in terms of accessibility.

https://adrianroselli.com/2019/03/under-engineered-toggles.html

> Use &lt;button> if:
> - you can count on JavaScript being available,
> - flipping the toggle has an immediate effect,
> - the toggle will never have an indeterminate state.

> Use &lt;input type="checkbox"> if:
> - you want to progressively enhance the control,
> - flipping the toggle will only take effect when the user submits it,
> - the toggle may have an indeterminate state.

By this logic and what the Switch component stands for, I think we should lean towards using 'button' instead of a 'checkbox' but it would be great if UX and engineering can give a heads-up on this before I go ahead and work further on it. 

According to [this section](https://adrianroselli.com/2019/08/under-engineered-toggles-too.html#Button), 

> If flipping the toggle immediately performs an action on the page (such as updating a setting), then a &lt;button> is a better fit. Unlike a checkbox, which only changes its state, a <button> conveys to a user that something will happen.
>
> If the control will never have an indeterminate state, then &lt;button> with aria-pressed may be right for you. It only accepts true or false and, unlike a checkbox, you cannot use script to set a third state.

<h3> &lt;button> with `aria-pressed` attribute v/s `role="switch"` </h3>

---
According to the official ARIA standards [1], 

> A switch provides approximately the same functionality as a checkbox and toggle button, but makes it possible for assistive technologies to present the widget in a fashion consistent with its on-screen appearance.

TL;DR - this may not be suitable for our need, though. 

Firstly, a checkbox control is the default fallback for a switch role _[1]_. Keeping in mind the above argument that a checkbox may not be what we want for our use-case, &lt;button> needs to be enhanced upon to have "switch" role and `aria-checked` attribute. However, that does not sound semantically right to me and from what I have looked up so far _[2][3]_, there is no great support for `role="switch"` across multiple ATs (especially with JAWS 2018 & 2019 + IE11). 

[1] https://www.w3.org/TR/wai-aria-1.1/#switch
[2] https://freedomscientific.github.io/VFO-standards-support/aria.html
[3] https://scottaohara.github.io/a11y_styled_form_controls/src/checkbox--switch/#affects_on_sr

The better alternative is to keep &lt;button> as the base element and use `aria-pressed` attribute for better screen-reader experience. 

Before I put out a technical design for this pattern, I think it'd be great if this is open for discussion and if UX and accessibility team can let me know if there is something more that I should consider.

--- 

**NOTE:** I have a [rough working code](https://github.com/cerner/terra-core/compare/nramamurth-switch) of Switch using checkbox and am currently working on getting it work using button as well.

cc / @neilpfeiffer @scottwilmarth @ryanthemanuel @ShettyAkarsh

**Comment by ShettyAkarsh** [comment]
@neilpfeiffer , @scottwilmarth request you guys to share your thoughts on the above findings.

**Comment by supreethmr** [comment]
> ### Switch as a <button> v/s <input type='checkbox'>
> I started off creating a prototype by implementing `Switch` as a "visually masked" checkbox and I had most of the accessibility concerns addressed except for one - whether or not the role should be "switch".
> 
> While researching further on that topic, I ran into the following post that talks about when it is appropriate to implement a "toggle" switch-like component as a `button` or `input type='checkbox'` with `role='switch'` and the implications of using "switch" as the role in terms of accessibility.
> 
> https://adrianroselli.com/2019/03/under-engineered-toggles.html
> 
> > Use <button> if:
> > 
> > * you can count on JavaScript being available,
> > * flipping the toggle has an immediate effect,
> > * the toggle will never have an indeterminate state.
> 
> > Use <input type="checkbox"> if:
> > 
> > * you want to progressively enhance the control,
> > * flipping the toggle will only take effect when the user submits it,
> > * the toggle may have an indeterminate state.
> 
> By this logic and what the Switch component stands for, I think we should lean towards using 'button' instead of a 'checkbox' but it would be great if UX and engineering can give a heads-up on this before I go ahead and work further on it.
> 
> According to [this section](https://adrianroselli.com/2019/08/under-engineered-toggles-too.html#Button),
> 
> > If flipping the toggle immediately performs an action on the page (such as updating a setting), then a <button> is a better fit. Unlike a checkbox, which only changes its state, a  conveys to a user that something will happen.
> > If the control will never have an indeterminate state, then <button> with aria-pressed may be right for you. It only accepts true or false and, unlike a checkbox, you cannot use script to set a third state.
> 
> ### <button> with `aria-pressed` attribute v/s `role="switch"`
> According to the official ARIA standards [1],
> 
> > A switch provides approximately the same functionality as a checkbox and toggle button, but makes it possible for assistive technologies to present the widget in a fashion consistent with its on-screen appearance.
> 
> TL;DR - this may not be suitable for our need, though.
> 
> Firstly, a checkbox control is the default fallback for a switch role _[1]_. Keeping in mind the above argument that a checkbox may not be what we want for our use-case, <button> needs to be enhanced upon to have "switch" role and `aria-checked` attribute. However, that does not sound semantically right to me and from what I have looked up so far _[2][3]_, there is no great support for `role="switch"` across multiple ATs (especially with JAWS 2018 & 2019 + IE11).
> 
> [1] https://www.w3.org/TR/wai-aria-1.1/#switch
> [2] https://freedomscientific.github.io/VFO-standards-support/aria.html
> [3] https://scottaohara.github.io/a11y_styled_form_controls/src/checkbox--switch/#affects_on_sr
> 
> The better alternative is to keep <button> as the base element and use `aria-pressed` attribute for better screen-reader experience.
> 
> Before I put out a technical design for this pattern, I think it'd be great if this is open for discussion and if UX and accessibility team can let me know if there is something more that I should consider.
> 
> **NOTE:** I have a [rough working code](https://github.com/cerner/terra-core/compare/nramamurth-switch) of Switch using checkbox and am currently working on getting it to work using button as well.
> 
> cc / @neilpfeiffer @scottwilmarth @ryanthemanuel @ShettyAkarsh

Hi, 

with reference of @nramamurth suggestion on using button instead of input with type `checkbox` in switch component, I have created [Draft PR](https://github.com/cerner/terra-core/pull/2816) and below are the links to component example and tests. 

Component : https://terra-core-deployed-pr-2816.herokuapp.com/components/terra-switch/switch/switch

Switch with Button : https://terra-core-deployed-pr-2816.herokuapp.com/tests/terra-switch/switch/default-switch-button

Switch with Input Type CheckBox : https://terra-core-deployed-pr-2816.herokuapp.com/tests/terra-switch/switch/default-switch-checkbox

I have added examples for both Switch with CheckBox and Switch with Button. Before proceeding further with Design, I would like to know if there is something more that I should consider.

cc / @neilpfeiffer @scottwilmarth @ryanthemanuel @ShettyAkarsh

**Comment by ShettyAkarsh** [comment]
@supreethmr ,  Can you post the design here for switch using button implementation ?

and also you could do the following changes.

1.  is the 'value' Prop required here using button implementation ?
2.  we could have keydown callback itself instead of handling everything through click.
3.  you can update the examples to work on space button click.
4.  And also should the switch slide should occur only when user enters 'Space' Key or even 'Enter'  Key as well ?

**Comment by scottwilmarth** [comment]
@supreethmr  apologies for being late to seeing this. I've tested both button and checkbox toggles on voiceover. Both seem to work very well. I have not had a chance to check to see how voice input would work with either of these types. I will ask Emily Ogle to test via JAWS. Even though, I'm a little late, here is another resource for reference regarding toggle and voice command: [Inclusive components (this link being toggle button) by Heydon Pickering.](https://inclusive-components.design/toggle-button/)

**Comment by supreethmr** [comment]
Thank you @scottwilmarth for sharing references. I found both button and checkbox are not being responsive to keyboard interactions in Jaws. will work on the changes by keeping above document as reference and will update the branch for review.

**Comment by supreethmr** [comment]
# Initial Tech Design for Switch
## Summary:
Switch component represents toggle button that indicates whether a setting is on or off and provide users with instantaneous feedback.

Will be using button instead of checkbox in switch component since checkbox with `role="switch"` has issues with JAWS. 

## Requirements:
[zeplin](https://zpl.io/aRP775E).

## Bidirectionality:
Alignment will be mirrored in rtl orientation.

## Accessibility:
Component will be keyboard navigatable
- Component can be focused with tab
- Pressing enter or space on the switch component will open the dropdown and focus the first option

## React Props:

| Prop | Type | Default | Description |
|----------|------|---------|--------------|
| isOn | boolean | false | Indicates Whether or not the Switch is enabled ("ON") |
| disabled | boolean | false | If true, the user won't be able to toggle the switch. |
| labelText | string | '' | Text of the label. |
| id | string | **required** | Id of the switch  |
| labelTextAttrs | object | {} | Additional attributes applied to the label. |
| onChange | function | none | Callback fired when the switch state is changed. |

## Example:
```javascript
<Switch
   id="test"
   isOn={value}
   labelText="Label"
   onChange={() => { setValue(!value); }}
/>
```

**Comment by supreethmr** [comment]
> @supreethmr , Can you post the design here for switch using button implementation ?
> 
> and also you could do the following changes.
> 
> 1. is the 'value' Prop required here using button implementation ?
> 2. we could have keydown callback itself instead of handling everything through click.
> 3. you can update the examples to work on space button click.
> 4. And also should the switch slide should occur only when user enters 'Space' Key or even 'Enter'  Key as well ?

I have posted tech design for switch component.
1. value prop is not required for button. it will be removed.
2. changed it onChange since onChange would be more appropriate for switch component.
3. fixed examples scrolling issue
4. it should slide on both enter and space key as per the [requirement](https://zpl.io/aRP775E)
